From be2833bd1e3baf621f5c2bad1cbf1815e6c096d0 Mon Sep 17 00:00:00 2001
From: Sachin Prabhu <sprabhu@redhat.com>
Date: Fri, 14 Nov 2014 16:45:41 -0500
Subject: [fs] Fix oops when creating symlinks on smb3

Message-id: <1415983541-5304-1-git-send-email-sprabhu@redhat.com>
Patchwork-id: 100054
O-Subject: [RHEL7.1 PATCH fs][bz1161429][SMB3] Fix oops when creating symlinks on smb3
Bugzilla: 1161429
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>

From: Steve French <smfrench@gmail.com>

We were not checking for symlink support properly for SMB2/SMB3
mounts so could oops when mounted with mfsymlinks when try
to create symlink when mfsymlinks on smb2/smb3 mounts

Signed-off-by: Steve French <smfrench@gmail.com>
Cc: <stable@vger.kernel.org> # 3.14+
CC: Sachin Prabhu <sprabhu@redhat.com>
(cherry picked from commit da80659d4aa758dc6935b10ec64513f0b67bc969)

Fixes bz 1161429. Tested successfully using kernel 3.10.0-203.el7.

Fixes a regression in RHEL 7.1.

Signed-off-by: Sachin Prabhu <sprabhu@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/cifs/link.c b/fs/cifs/link.c
index 264ece7..6b532a6 100644
--- a/fs/cifs/link.c
+++ b/fs/cifs/link.c
@@ -213,8 +213,12 @@ create_mf_symlink(const unsigned int xid, struct cifs_tcon *tcon,
  if (rc)
   goto out;
 
- rc = tcon->ses->server->ops->create_mf_symlink(xid, tcon, cifs_sb,
-     fromName, buf, &bytes_written);
+ if (tcon->ses->server->ops->create_mf_symlink)
+  rc = tcon->ses->server->ops->create_mf_symlink(xid, tcon,
+     cifs_sb, fromName, buf, &bytes_written);
+ else
+  rc = -EOPNOTSUPP;
+
  if (rc)
   goto out;
 
-- 
1.7.1